src/relay/blockchain/events_informations.py

```diff
@@ -217,12 +217,20 @@ def balance_viewed_from_user(balance_update_event):
 
 
 def sorted_events(events, reverse=False):
-    def key(event):
+    def log_index_key(event):
+        if event.log_index is None:
+            raise RuntimeError("No log index, events cannot be ordered truthfully.")
+        return event.log_index
+
+    def block_number_key(event):
         if event.blocknumber is None:
             return math.inf
         return event.blocknumber
 
-    return sorted(events, key=key, reverse=reverse)
+    # We sort first by index, then by block number
+    # the second sort will keep the events ordered by log_index within each block
+    events_sorted_by_index = sorted(events, key=log_index_key, reverse=reverse)
```

Comment:
I dont think this is guarantied.
The correct way to do it is to have key return a tuple `(block_number, log_index)`
Tuples are compared first by first element then by second, and so on.